syntax = "proto3";
package video_call;

//errors response
message ErrorRes {
    int64 code = 1;
    string message = 2;
}

message BaseResponse {
    bool success = 1;
    ErrorRes errors = 2;
}

message ServerResponse {
    StunServer stun_server = 1;
    TurnServer turn_server = 2;
    string group_rtc_token = 3;
    VideoCallMessage video_call = 4;
}

message VideoCallMessage {
  string id = 1;
  string message_id = 2;
  string call_status = 3;
  string call_type = 4;
  int64 created_at = 5;
  int64 started_at = 6;
  int64 ended_at = 7;
}

message StunServer {
  string server = 1;
  int64 port = 2;
}

message TurnServer {
  string server = 1;
  int64 port = 2;
  string type = 3;
  string user = 4;
  string pwd = 5;
}

// Request: new call
message VideoCallRequest {
  string client_id = 1;
  int64 group_id = 2;
  /* string call_type = 3; //audio or video */
  string call_id = 3;
}
//response: BaseResponse


// Request: call update
message UpdateCallRequest {
  int64 group_id = 1;
  string update_type = 2;
}
//response: BaseResponse

message VideoCallResponse {
  int64 group_id = 1;
  string from_client_id = 2;
  string client_id = 3;
  VideoCallMessage video_call = 4;
}

service VideoCall {
    rpc video_call(VideoCallRequest) returns (ServerResponse) {};
    rpc cancel_request_call(VideoCallRequest) returns (VideoCallResponse) {};
    rpc miss_call(VideoCallRequest) returns (VideoCallResponse) {};
    rpc decline_call(VideoCallRequest) returns (VideoCallResponse) {};
    rpc end_call(VideoCallRequest) returns (VideoCallResponse) {};
    rpc update_call(UpdateCallRequest) returns (VideoCallResponse) {};
}
